### ISSUE TITLE
 frame.js does not work on LegacyFox+VimFx

### ISSUE THREAD

**Issue by azuwis** [akhodakivskiy/VimFx#939]
It seems [frame.js](https://github.com/akhodakivskiy/VimFx/blob/master/documentation/config-file.md#framejs) does not work on LegacyFox+VimFx.

For example, [Search for selected text](https://github.com/akhodakivskiy/VimFx/wiki/Custom-Commands#search-for-the-selected-text) does not work.

No errors in browser console, just warnings:
```
Use of nsIFile in content process is deprecated. osfile_async_front.jsm:80:29
```

OS: Debian Unstable and Windows 10
Firefox: 69.0
VimFx: 0.22.0

Is it a known limitation?

@girst Thanks for your work, I'm so happy to use VimFx again.

**Comment by girst** [comment]
On Sun, Sep 22, 2019 at 08:40:05PM -0700, Zhong Jianxin wrote:
>It seems [frame.js] does not work on LegacyFox+VimFx.

what browser version and LegacyFox commit are you using?

>Use of nsIFile in content process is deprecated.  osfile_async_front.jsm:80:29
>Is it a known limitation?

I'll look into it.

>@girst Thanks for your work, I'm so happy to use VimFx again.

nice to know that I'm not screaming/developing into the void ;)

**Comment by azuwis** [comment]
OS: Debian Unstable and Windows 10
Firefox: 69.0 https://packages.debian.org/sid/firefox
LegacyFox: Current master https://github.com/girst/LegacyFox/commit/d1c80010b1fb1e352fb8754ace534e8dbe8cf954
VimFx: 0.22.0

**Comment by azuwis** [comment]
When reloading config using `gC` shortcut, the following error appears:
```
VimFx: Error loading frame.js file:///home/azuwis/.vimfx/frame.js Error opening input stream (invalid filename?): file:///home/azuwis/.vimfx/frame.js 3 config.js:53:13
```

File `/home/azuwis/.vimfx/frame.js` does exist, and worked on Firefox 56 + VimFx.

**Comment by girst** [comment]
OK, I've done some digging: the error about invalid file URIs you're 
getting is a sandbox feature: Frame scripts are not allowed to access 
the file system (through file://) any more.

Here are some documents I've found for my own future reference:
- https://wiki.mozilla.org/Security/Sandbox
- https://bugzilla.mozilla.org/show_bug.cgi?id=1288874
- https://bugzilla.mozilla.org/show_bug.cgi?id=1221148

Would you mind trying your frame script with a workaround?  In 
about:config, search for security.sandbox.content.level and set it to 2
(note that the sandbox levels are platform-dependent; on windows you 
might need to lower it even further, but I don't have a windows box to 
test).
/!\ But big heads-up: this is a security risk, as it will weaken the 
sandbox.  so please reenable it after testing if everything works and 
report back, then. I've got an idea how to fix this without compromising 
security.

For the proper fix, I'm planning to use chrome.manifest overrides 
installing a resource:// or chrome:// URI for config.js and frame.js and 
pointing them to wherever the user has those files.

**Comment by girst** [comment]
*tl;dr*: there are some workarounds we can safely use:
1. Linux: security.sandbox.content.read_path_whitelist
2. OSX: $PROFILE/{extensions,chrome}
3. Windows: no idea --  @azuwis, please test if any of the above 
workarounds work, or what security.sandbox.content.level allows 
read-access to common locations (restart browser after changing pref).

those will be documented, but need testing beforehand. anyone got a 
windows or osx box?

 - - - - -

Following are some notes mostly for myself:

OK, the chrome.manifest override path was a dead-end (i thought it was 
possible to access either resource:// URIs or chrome:// URIs marked 
contentaccessible=yes, but neither works). 

Greasemonkey ran into that issue back in 2017[1][2], and they chose to 
send an IPC message from (in our terms) frame script to the chrome 
script, which slurps the file and sends it as another IPC message, and 
finally call Cu.evalInSandbox() from frame context.

[1]: https://github.com/greasemonkey/greasemonkey/issues/2485
[2]: https://github.com/greasemonkey/greasemonkey/pull/2486/files

AdBlockPlus decided[3] to create a new about: page, as that apparently 
can be accessed by content scripts, sometimes[4]. Mozilla has some 
example code[5].

[3]: https://github.com/adblockplus/adblockplus/blob/8ecfee137916f70d956e1207048d4b6ef8873a26/lib/child/elemHide.js#L50
[4]: https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Multiprocess_Firefox/Which_URIs_load_where
[5]: https://developer.mozilla.org/en-US/docs/Archive/Add-ons/Custom_about:_URLs

Both of these workarounds are hacks I don't really want to implement and 
test, because I'm uncertain about what kind of edge cases they entail.

There's also moz-extension://, but I haven't looked into if that could 
work (this random bug might do something kinda what we want: 
https://bugzilla.mozilla.org/show_bug.cgi?id=1334550; keeping in mind 
that this protocol was designed for webextensions).

 - - - - -

I've did look more into Firefox' sandbox config, and extracted the 
following (I'll write this down in VimFx' docs, once we have info on 
windows and possibly osx):

1. on Linux (others untested--please report), one can safely exclude 
files from being blocked by setting 
security.sandbox.content.read_path_whitelist to a comma-seperated list 
of filenames[6], possibly also directory names.

example user.js:
	user_pref("security.sandbox.content.read_path_whitelist", "/home/girst/.vimfx/config.js, /home/girst/.vimfx/frame.js");

2. on OSX it looks like moving .vimfx to $PROFILE/{extensions,chrome} 
should work (anyone care to test this?) [7].

3. on other platforms, one could at least fall back to
	user_pref("security.sandbox.content.level", 1);
(again: a security risk if your content process gets owned[8]; a value 
of 2 works on linux too)

The windows sandbox is ... well, I have no idea what that[9][10] does.


[6]: https://searchfox.org/mozilla-central/rev/45f30e1d19bde27bf07e47a0a5dd0962dd27ba18/ipc/glue/GeckoChildProcessHost.cpp#536
[7]: https://searchfox.org/mozilla-central/rev/45f30e1d19bde27bf07e47a0a5dd0962dd27ba18/browser/app/profile/firefox.js#1091
[8]: https://bugzilla.mozilla.org/show_bug.cgi?id=1221148#c30
[9]: https://searchfox.org/mozilla-central/rev/45f30e1d19bde27bf07e47a0a5dd0962dd27ba18/security/sandbox/win/src/sandboxbroker/sandboxBroker.cpp#428
[10]: https://searchfox.org/mozilla-central/source/security/sandbox/chromium/sandbox/win/src/security_level.h

**Comment by azuwis** [comment]
Thanks for the analysis.

Here are my test results:

OS | security.sandbox.content.level | security.sandbox.content.read_path_whitelist | works
-- | ------------------------------ | --------------------------------------------- | -----
Linux | 2 | reset | yes
Linux | 3 | reset | no
Linux | reset | /home/azuwis/.vimfx/frame.js | yes
Linux | reset | file:///home/azuwis/.vimfx/frame.js | no
Windows | 2 | reset | yes
Windows | 3 | reset | no
Windows | reset | C:\Users\azuwis\.vimfx\frame.js | yes!

**Comment by azuwis** [comment]
I put this in `config.js` as a workaround, it works on both Linux and Windows:

``` js
Cu.import('resource://gre/modules/XPCOMUtils.jsm')
XPCOMUtils.defineLazyModuleGetter(this, 'Preferences', 'resource://gre/modules/Preferences.jsm')
Preferences.set('security.sandbox.content.read_path_whitelist', OS.Path.fromFileURI(`${__dirname}/frame.js`))
```

But need to restart Firefox twice at the first time for obvious reason, not a perfect solution.

**Comment by girst** [comment]
On Tue, Sep 24, 2019 at 11:14:55PM -0700, Zhong Jianxin wrote:
>Here are my test results:

Very thorough, thanks!

I've read some more into the whitelist, and it appears to be not 
available on OSX, where there are instead two config flags that can hold 
1 directory name each[1]. Judging by that comment[2], Mozilla really 
doesn't want us to depend on that, so I'll probably mention the 
(insecure) s.s.content.level=2 workaround in the docs as well for future 
reference (with a big warning/explanation, of course).

[1]: https://searchfox.org/mozilla-central/rev/f43ae7e1c43a4a940b658381157a6ea6c5a185c1/testing/mozbase/mozprofile/mozprofile/profile.py#228
[2]: https://searchfox.org/mozilla-central/rev/f43ae7e1c43a4a940b658381157a6ea6c5a185c1/dom/ipc/ContentChild.cpp#1705

I think the plan forward is to:

1. update the docs to require setting
   security.sandbox.content.read_path_whitelist (or the osx equivalent)
2. provide a better error message than the useless one from the browser
   in case the user forgets/isn't aware of that change.

I'll do both as time allows, hopefully this week.

regarding your other comment: I don't see the value in that workaround.  
you'll have to set extensions.vimfx.config_path (or whatever it's 
called) anyways, so why not just use a $PROFILE/user.js file to set both 
at once?

**Comment by azuwis** [comment]
The workaround is for Windows, with the workaround I can use the same settings for Linux and Windows.

`$PROFILE/user.js`:
```
user_pref("extensions.VimFx.config_file_directory", "~/.vimfx");
```

`$VIMFX_CONFIG/config.js`:
```
Cu.import('resource://gre/modules/XPCOMUtils.jsm')
XPCOMUtils.defineLazyModuleGetter(this, 'Preferences', 'resource://gre/modules/Preferences.jsm')
Preferences.set('security.sandbox.content.read_path_whitelist', OS.Path.fromFileURI(`${__dirname}/frame.js`))
```

Without the workaround, I need to set this in Linux:
```
user_pref("security.sandbox.content.read_path_whitelist", "/home/azuwis/.vimfx/frame.js");
```

And this in Windows:
```
user_pref("security.sandbox.content.read_path_whitelist", "C:\Users\azuwis\.vimfx\frame.js");
```

Is there a way to do some logic in `user.js`, Like setting different value base on OS?

**Comment by azuwis** [comment]
[Search for selected text](https://github.com/akhodakivskiy/VimFx/wiki/Custom-Commands#search-for-the-selected-text) still does not work, because `window.BrowserSearch.loadSearch()` function is gone.

Did some research, but only found webextention API `browser.search.search()`.

Is there a way to use webextension API or other way to do browser search?

**Comment by girst** [comment]
I've written up the documentation and added a more helpful error 
message. Would you mind reviewing that, please? It is pushed to my repo 
at https://github.com/akhodakivskiy/VimFx/commit/711aafb4f968090f1c3d167d283fc2abe0d97515 

The new error message kicks in when loading frame.js fails and we detect 
that the sandbox is not configured for VimFx. It reads:
> VimFx: Error loading #{file} likely due to e10s sandbox
> Please consult VimFx' documentation on config files.

regarding your workaround: oh, i get it now! Sadly, there isn't a way to 
do custom logic in user.js (despite its name it is not interpreted by 
the Javascript engine any more[1]). If you don't mind keeping patches on 
top of LegacyFox, you can probably put it there like so (untested):

    const {Services} = Cu.import('resource://gre/modules/Services.jsm');
    if (Services.appinfo.OS == "WINNT") {
      // valid values: 'Darwin', 'Linux', and 'WINNT'
      pref("security.sandbox.content.read_path_whitelist", "C:\\whatever\\...");
    } else {
      pref("security.sandbox.content.read_path_whitelist", "/home/whatever/...");
    }

I don't think you can reliably read out the value of ~/.vimfx from 
there, as (i believe) autoconfig scripts are loaded before user 
preferences.

[1]: https://blog.mozilla.org/nnethercote/2018/03/09/a-new-preferences-parser-for-firefox/

regarding the 'search for selected text' bug: that's a seperate issue; 
please open a new ticket for that.

**Comment by girst** [comment]
sorry for the bugspam, i've had to update the fix. here's the latest 
commit URL: 
https://github.com/girst/VimFx/commit/6daba8c829fd54e36e97079dd61da8245579f60b

**Comment by azuwis** [comment]
Built the XPI and tested it.

The error message does appear on both Linux and Windows:
```
VimFx: Error loading frame.js likely due to e10s sandbox config.js:72:15
    loadFile chrome://vimfx/content/lib/config.js:72
    loadConfig chrome://vimfx/content/lib/main-frame.js:48
    exports chrome://vimfx/content/lib/main-frame.js:51
    startup chrome://vimfx/content/bootstrap.js:75
    <anonymous> chrome://vimfx/content/bootstrap.js:104
    invokeListener chrome://vimfx/content/lib/message-manager.js:114
    fn chrome://vimfx/content/lib/message-manager.js:63
Please consult VimFx' documentation on config files. config.js:73:15
```

And when `security.sandbox.content.read_path_whitelist` or
`security.sandbox.content.level` are set, no error message.

Another finding, `security.sandbox.content.read_path_whitelist` does not need
restarting Firefox to take effect, a new tab will do.

**Comment by girst** [comment]
On Fri, Sep 27, 2019 at 01:15:38AM -0700, Zhong Jianxin wrote:
>Another finding, `security.sandbox.content.read_path_whitelist` does 
>not need
>restarting Firefox to take effect, a new tab will do.

Great! I'll update that documentation and tag a new release! Thanks for 
all your help!

**Comment by girst** [comment]
Fix released!
